Support setting outgoing interface for multicast on Socket

tmds opened the pull request:
Fix https://github.com/dotnet/corefx/issues/6486

pal_networking: support specifying the network interface for outgoing multicast traffic via SocketOptionName.MulticastInterface on non-Windows platforms

The interface for outgoing multicast traffic can be set as follows:
	socket.SetSocketOption(SocketOptionLevel.IP, SocketOptionName.MulticastInterface, IPAddress.HostToNetworkOrder(interfaceIndex))

stephentoub commented:
cc: @ericeil, @pgavlin, @davidsh

ericeil commented on src/Native/System.Native/pal_networking.cpp:
```diff
@@ -2049,6 +2049,27 @@ SystemNative_SetSockOpt(int32_t socket, int32_t socketOptionLevel, int32_t socke
         return PAL_ENOTSUP;
     }
 
+#ifndef WIN32
```
I don't think this `#ifndef` is necessary.

stephentoub replied:
It's not.  This library only compiles for Unix.

ericeil commented:
Do we need to also do the reverse translation for `getsockopt`?
Can you add a couple of test cases for getting/setting this option?

stephentoub commented on src/Native/System.Native/pal_networking.cpp:
```diff
@@ -2049,6 +2049,27 @@ SystemNative_SetSockOpt(int32_t socket, int32_t socketOptionLevel, int32_t socke
+        }
+    }
+#endif /* WIN32 */
+
```
I'm wondering if this translation would be better to do in the C# code.  Essentially the C# code would do the translation (maybe here https://github.com/dotnet/corefx/blob/master/src/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Unix.cs#L966) and then call Interop.Sys.IPv4MulticastOption, either directly or via a call to SetMulticastOption (which would need to be tweaked to allow this option).  It'd be nice to be able to keep this native function as a simple wrapper around setsockopt.

tmds replied:
Makes sense.
It would be even nicer if SetMulticastOption would accept a MulticastInterface because the API to set the interface index would be cleaner. Off course, it needs to work on all platforms.
For now I won't touch SetMulticastOption but I'll move the translation to SetSockOpt and call Interop.Sys.SetIPv4MulticastOption from there.

stephentoub commented:
@tmds, by the way, thanks for submitting the PR!  It's great when someone opens an issue and then submits a PR to fix it! :+1:

tmds commented:
@ericeil I will add some tests. Currently I don't intend to implement the getter.

ericeil commented:
> Currently I don't intend to implement the getter.

I would not want to have the setter without the corresponding getter.  I'm not sure whether that means we need the getter before we merge this, or maybe it's good enough to just file another issue prior to merging.  @stephentoub, do you have an opinion here?

stephentoub commented:
> @stephentoub, do you have an opinion here?

We should have the getter as well, both to support roundtripping with the setter but also to be consistent with the Windows implementation and its support.  It'd be great if it could be done at the same time.  If that's not possible, then an issue and a follow-up PR to handle it would be ok.

tmds commented:
@stephentoub @ericeil Just learned this myself: it isn't possible to read back the multicast interface index on linux, [setsockopt](https://github.com/torvalds/linux/blob/dea08e604408d0303e2332896c5fdd8c1f7d79a2/net/ipv4/ip_sockglue.c#L805) accepts an in_addr, ip_mreq, ip_mreqn. [getsockopt](https://github.com/torvalds/linux/blob/dea08e604408d0303e2332896c5fdd8c1f7d79a2/net/ipv4/ip_sockglue.c#L1388) on the other hand, only accepts an in_addr.

Should the value be stored in a field?

ericeil commented:
> Should the value be stored in a field?

That seems like it's probably the only option.  That makes @stephentoub's suggestion of moving the translation to managed code even more appealing....

stephentoub commented:
Alternatively, if the platform really doesn't support it (I haven't investigated), we could throw PlatformNotSupportedException.

tmds commented:
@stephentoub what is the appropriate way to run a specific test project using ./run-test.sh? What should be done before to make sure the native and required managed parts are compiled?

stephentoub commented:
> what is the appropriate way to run a specific test project using ./run-test.sh? What should be done before to make sure the native and required managed parts are compiled?

To be honest, I'm not sure, as I don't typically use it.  @ellismg, do you know?  I'm still using a workflow from before this existed, where I have a directory that contains all of my binaries, and I just put new versions there and run them directly with ```corerun xunit.console.netcore.exe ...``` etc.  Gives me a bit more control.

stephentoub commented on src/System.Net.Sockets/tests/FunctionalTests/SocketOptionNameTest.cs:
```diff
@@ -95,5 +95,39 @@ public void MulticastOption_CreateSocketSetGetOption_GroupAndInterfaceIndex_SetS
+            {
+                s.SetSocketOption(SocketOptionLevel.IP, SocketOptionName.MulticastInterface, IPAddress.HostToNetworkOrder(index));
+            }
+        }
```
This test and the one above it can be condensed into a single [Theory], e.g.
```C#
[Theory]
[InlineData(0)]
[InlineData(1)]
public void MulticastInterface_SetToInterface_ValidIndex_Succeeds(int index)
{
    ...
}
```

stephentoub commented on src/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Unix.cs:
```diff
@@ -963,6 +963,26 @@ public static unsafe SocketError SetSockOpt(SafeCloseSocket handle, SocketOption
+                if (optionName == SocketOptionName.MulticastInterface)
+                {
+                    /* if the value of the IP_MULTICAST_IF is an address in the 0.x.x.x block
+                       the value is interpreted as an interface index */
```
Nit: I think elsewhere we've shied away from using ```/*...*/``` syntax across multiple lines, so maybe just use ```//``` syntax.

stephentoub commented on src/System.Net.Sockets/tests/FunctionalTests/SocketOptionNameTest.cs:
```diff
@@ -95,5 +95,39 @@ public void MulticastOption_CreateSocketSetGetOption_GroupAndInterfaceIndex_SetS
+                Assert.Throws<SocketException>(() =>
+                    s.SetSocketOption(SocketOptionLevel.IP, SocketOptionName.MulticastInterface, IPAddress.HostToNetworkOrder(index)));
+            }
+        }
```
Is there any test we could write to validate that SetSocketOption not only returned successfully but also that that the option had the intended affect on behavior?

stephentoub commented:
I like this version much better.  A few small nits, but otherwise LGTM.  Thanks for fixing this. (And we'll need to follow-up on what we want to do for the getsockopt side of things.)

ellismg commented:
>> what is the appropriate way to run a specific test project using ./run-test.sh? What should be done before to make sure the native and required managed parts are compiled?

>To be honest, I'm not sure, as I don't typically use it. @ellismg, do you know? I'm still using a workflow from before this existed, where I have a directory that contains all of my binaries, and I just put new versions there and run them directly with corerun xunit.console.netcore.exe ... etc. Gives me a bit more control.

There is a [-restrict-proj](https://github.com/dotnet/corefx/blob/master/run-test.sh#L41) option you can pass to `run-test.sh` which will control what test projects it runs.  Every time you run it, it will redeploy all the native bits (both coreclr and corefx), the managed bits remain unchanged so if you have made test or product fixes you have to copy them into the test folders manually.

tmds commented:
@ellismg I had to specify --coreclr-bins and --mscorlib-bins to make run-test.sh work. The script has some logic to figure out these paths. Perhaps this is broken.

tmds commented:
@stephentoub I rewrote the test so it is sending a message and receiving one. Either test is fine by me, the new version is more likely to fail occasionally because it interacts more with the underlying OS (e.g. firewal) and network interaction.

The first version was more minimal and it covered a case which Threw before the fix.
Before the fix, linux would search for a network interface with the IP address 0.0.0.1, which failed. Now it is correctly using the interface with index 1.

stephentoub commented on src/System.Net.Sockets/tests/FunctionalTests/SocketOptionNameTest.cs:
```diff
@@ -95,5 +96,53 @@ public void MulticastOption_CreateSocketSetGetOption_GroupAndInterfaceIndex_SetS
+        public void MulticastInterface_Set_ValidIndex_Succeeds(int interfaceIndex)
+        {
+            IPAddress multicastAddress = IPAddress.Parse("239.1.2.3");
+            int port = 54321;
```
Can this instead bind to any available port?  It's getting "address in use" errors on OS X, and I'm assuming this is the cause.  Same for the other tests.

stephentoub commented:
@ericeil, do you have a preference on the tests?

@tmds, thanks for adding these.  I assume one problem with the new version of the tests is that occasionally we may lose one of the UDP packets, causing the Receive to hang?  It'd be good to instead timeout the receive after a period of time (there should be examples in some other tests.  And if we expect it to fail occasionally, it might be good to send multiple packets (but still only receive one), enough to counteract any reasonable chance that they'd all get lost.

stephentoub commented on src/System.Net.Sockets/tests/FunctionalTests/SocketOptionNameTest.cs:
```diff
@@ -95,5 +96,53 @@ public void MulticastOption_CreateSocketSetGetOption_GroupAndInterfaceIndex_SetS
+        public void MulticastInterface_Set_ValidIndex_Succeeds(int interfaceIndex)
+        {
+            IPAddress multicastAddress = IPAddress.Parse("239.1.2.3");
+            int port = 54330 + interfaceIndex;
```
Can you use 0 and then access the LocalEndPoint after the bind to determine which port it bound to?  When we've hardcoded ports in tests in the past, it's led to problems due to conflicts with other things on the machine.

tmds commented:
@stephentoub To make the OS assign a port, the socket needs to do a send and that what I added. I separated it into it's own method because it seems really out of place in the test method.
I also added the receive timeout. Because these packets are sent and received on the same host, I don't expect packets to get lost. In case my assumption is wrong, the receive will now throw on timeout and then we can improve the test by sending the packet several times.

tmds commented:
@stephentoub Seems to work fine everywhere except OSX. I don't have an OS X system to debug the issue.
Can you do another review of the code?

ericeil commented:
According to the [OSX docs](https://developer.apple.com/library/mac/documentation/Darwin/Reference/ManPages/man4/ip.4.html), `IP_MULTICAST_IF` accepts a `struct in_addr`.  If I'm reading this change correctly, you're passing an `ip_mreq` instead.  Also, the address passed to `IP_MULTICAST_IF` looks like it needs to be the actual address of the interface, not the index.  Perhaps that explains why this isn't passing on OSX?

(I'm not sure if the above is also true on Linux or not.)

tmds commented:
@ericeil This patch is adding support for setting the multicast interface by index instead of by address. The initial version of the test, which only did a set of the option, passed on all systems.

The test fails on the command where it is supposed to receive a message which was sent earlier.
The log also only shows a fail on the interfaceIndex: 1 (I assume this to be loopback) but not on the interfaceIndex: 0 (any).

This should be debugged on a Mac system. These are somethings which can be checked:
* Does the binding to a 'free' port work okay? receiveSocket.LocalEndPoint should be '0.0.0.0' at this point: https://github.com/dotnet/corefx/pull/6523/files#diff-19adc45fa233056f6babaef13639a1acR145
* Does enabling loopback have any effect? Try setting MulticastLoopback to true on both the receive and send socket at https://github.com/dotnet/corefx/pull/6523/files#diff-19adc45fa233056f6babaef13639a1acR106
* Is there anything particular about multicast on loopback of OS X?

If someone can debug this that would be nice. Otherwise, perhaps we should disable the test for OSX?

ericeil commented:
As I said before, I'm not sure OSX supports passing `ip_mreq` to the `IP_MULTICAST_IF` socket option.  I see that Linux does explicitly support this, but here is what the [OSX man page](https://developer.apple.com/library/mac/documentation/Darwin/Reference/ManPages/man4/ip.4.html) has to say:

> For hosts with multiple interfaces, each multicast transmission is sent from the primary network interface.  The IP_MULTICAST_IF option overrides the default for subsequent transmissions from a given socket:
>
>     struct in_addr addr;
>     setsockopt(s, IPPROTO_IP, IP_MULTICAST_IF, &addr, sizeof(addr));
>
>  where "addr" is the local IP address of the desired interface or INADDR_ANY to specify the default interface.  An interface's local IP address and multicast capability can be obtained via the SIOCGIFCONF and SIOCGIFFLAGS ioctls.  Normal applications should not need to use this option.

I don't see anything in there about an option to pass `ip_mreq` instead.  So it looks to me like there is additional work to make this happen on OSX.  

I can try to look at this under a debugger later today; maybe this actually does work, and something else is going wrong.

ericeil commented:
BTW I wouldn't mind if this threw `PlatformNotSupportedException` on OSX (if the caller specifies an index, not an address) for now, with an issue filed to find a way to make this work in the future.

tmds commented:
@ericeil The code to handle ip_mreqn on OSX is here: https://github.com/opensource-apple/xnu/blob/977178fc3e01a280624e2095838cb9530763cda9/bsd/netinet/in_mcast.c#L2625

If you don't have time to investigate now, throwing PlatformNotSupportedException is also fine by me.
Do you have an example of checking for OSX and throwing PlatformNotSupportedException?

ericeil commented:
I spent a little time debugging this.  The test passes on my Mac if I use a different interface index.  Index 1 is "loopback," as expected, but I only receive the packets if I send them from interface 4 ("en0").  I assume this is because the receiving socket got bound to interface 4, so it doesn't see the packets sent to interface 1, but I don't know how to confirm that directly.

Setting `MulticastLoopback` did not seem to help.

I wonder if it's possible to ensure that the receiving socket is bound to the interface from which we're sending?

In any case, I think I'd call this a test issue at this point, since everything does appear to work if configured correctly for the platform.  I'd disable the test on OSX, and we can follow up on fixing the test with a separate issue.

Also, I'm a little concerned that we're taking advantage of an essentially undocumented feature of OSX.  We should at least note that in the code comments.

tmds commented:
Thanks for investigating further.

Perhaps the LocalAddress of the sockets will tell you something. The receiveSocket should have an address of 0.0.0.0 with a port at this point https://github.com/dotnet/corefx/pull/6523/files#diff-19adc45fa233056f6babaef13639a1acR145.
For debugging, you can replace the send part in CreateBoundUdpSocket with a Bind on 127.0.0.1 with a free port.

As for the documentation part, the OSX kernel is based on the BSD kernel which has exactly the same code: https://github.com/freebsd/freebsd/blob/2bb0e924cc5d81b56b5eba9cae7011c55a7c6215/sys/netinet/in_mcast.c#L2454. Here the feature is documented: https://github.com/freebsd/freebsd/blob/0f2211c30ded7871b441bfaa1647843f1b4fd0c2/share/man/man4/ip.4#L479.

I'll disable the test for OSX.

ericeil commented:
Filed #6781 for the CentOS test failure.

ericeil commented:
@dotnet-bot test Innerloop CentOS7.1 Release Build and Test please

ericeil commented:
@tmds, this looks good to me.  Please squash your commits, and then I can merge.

ericeil commented:
Thank you for the contribution!

tmds commented:
No problem. Thanks for your feedback.
